core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -1374,11 +1374,35 @@ private Result toInnerStorageType(Result result, Type storageType) {
     }
     final Type storageType = currentStorageType != null
         ? currentStorageType : typeFactory.getJavaClass(dynamicParam.getType());
-    final Expression valueExpression =
+
+    // For numeric types, get the value using the following functions on the
```

Comment:
This may work for numeric values, but does not seem to be a general solution for other types. Is there a design that would handle that? It looks like any source/destination type combination is possible. So perhaps this could be handled like a cast expression? Although I don't see the destination type in this function.

Reply:
Indeed. The PR only fixes the issue for numeric values, but for other types the user would still encounter a `ClassCastException`. Ideally, we would have a function that handles all implicit conversions and fails with a user-friendly error for unsupported conversions. I am sure this functionality already exists in the Calcite in the form of `CAST` expressions. Possibly it could be leveraged here.

Reply:
As @mihaibudiu suggests, this is basically a cast, which is already available via the `Types.castIfNecessary` method, see below for how it can be used:

```
    final boolean isNumeric = SqlTypeFamily.NUMERIC.contains(dynamicParam.getType());
    final Expression valueExpression = Types.castIfNecessary(storageType,
        EnumUtils.convert(Expressions.call(root, BuiltInMethod.DATA_CONTEXT_GET.method,
                Expressions.constant("?" + dynamicParam.getIndex())),
            isNumeric ? java.lang.Number.class : storageType));
```

If you replace your code with this, it passes all the tests you added, and it might be working for non-numeric cases too (maybe with some adaptations).

Could you give it a try and add more tests non-numeric types to check?

Reply:
Thanks for the suggestion! It greatly simplifies the fix.

I made the requested change, but kept the null check as the cast operation may fail. For instance, `o.intValue()` requires `o` to be non-null. Otherwise, `testPreparedStatement` would fail.

As for additional tests, do you have in mind assigning a non-numeric value to a numeric value? These would still raise an exception:

```java
  @Test void bindStringParameter() {
    for (SqlTypeName tpe : SqlTypeName.INT_TYPES) {
      final String sql =
          "with cte as (select cast(100 as " + tpe.getName() + ") as empid)"
              + "select * from cte where empid = ?";

      CalciteAssert.hr()
          .query(sql)
          .consumesPreparedStatement(p -> {
            p.setString(1, "100");
          })
          .returnsUnordered("EMPID=100");
    }
  }
```

This throws a `ClassCastException` which we could rewrite into a more user-friendly message. Alternatively, we could prevent the conversion happening in the first place by introducing a validation check.

```
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Number
```

Reply:
I personally would file a new Jira issue for the casts not covered and merge this PR as is.
You have to decide which casts can appear in this context.

Reply:
Agreed. It seems unrelated to this PR. I created a ticket for it: https://issues.apache.org/jira/browse/CALCITE-6284

Reply:
Sorry I haven't been very precise, I didn't mean to cast Strings to numbers, what I had in mind was values from the string family to be used in parameters of the same family, possibly with (e.g., `varchar(2)` value for `varchar(3)` parameter), but I don't know if it is even expressible with the APIs we use to pass literals in, so feel free to leave this outside the scope of the present PR in case.

Reply:
> I made the requested change, but kept the null check as the cast operation may fail. For instance, `o.intValue()` requires `o` to be non-null. Otherwise, `testPreparedStatement` would fail.

I might have made a mistake but I tried all your tests with the suggested piece of code and all passed.
Did you add a new test specifically for this maybe?

Reply:
Yes, that makes sense. I added a similar test based on Mihai's request below. Since this PR is mostly about numeric types, `VARCHAR` conversions could be tackled in a separate ticket.

Indeed. All the tests I added were passing with your change, but `testPreparedStatement` existed before and started failing. Explicitly checking whether the value is `null`, fixed it.